capsules/src/net/thread/tlv.rs

```diff
@@ -0,0 +1,1315 @@
+//! Implements Thread 1.1.1 Specification Type-Length-Value (TLV) encoding
+//! and decoding. Used during mesh link establishment (MLE), outlined in
+//! Chapter 4.
+//!
+//! To simplify variable-length TLV value decoding in Rust, values are
+//! assumed to have a maximum length of 128 bytes. This assumption is made
+//! only when a variable-length value must be decoded from network byte
+//! order before it can be interpreted correctly. Excluded from this case
+//! are variable-length values that contain data that must later be decoded
+//! by the caller before being interpreted (for example, sub-TLVs). These
+//! values are instead returned as a slice of the original buffer passed to
+//! the decode function.
+//!
+//! This module currrently implements the minimum subset of TLVs required
+//! to connect a Sleepy End Device (SED) to a Thread network.
+//!
+//! Author: Mateo Garcia
+//!         mateog@stanford.edu
+
+
+// NOTES FOR DEBUGGING:
+// - .to_be() may not have been called on values wider than one byte
+// - encode_bytes_be may have been used instead of encode_bytes
+// - decode_bytes_be may have been used instead of decode_bytes
+// - See 4.5.25 Active Operational Dataset TLV and 4.5.26 Pending Operational Dataset TLV
+//    - Are Active and Pending Timestamp TLVs, respectively, required to be sent as well
+//      if either of the dataset tlvs are sent?
+
+
+use core::mem;
+use net::stream::{decode_u8, decode_u16, decode_u32, decode_bytes_be};
+use net::stream::{encode_u8, encode_u16, encode_u32, encode_bytes, encode_bytes_be};
+use net::stream::SResult;
+
+const TL_WIDTH: usize = 2; // Type and length fields of TLV are each one byte.
+const MAX_VALUE_LENGTH: usize = 128; // Assume a TLV value will be no longer than 128 bytes.
+
+pub enum Tlv<'a> {
```

Comment:
All of the public facing types and methods need documentation. Also, note that none of these variants are marked `pub`, which means a client module wouldn't be able to construct a `Tlv` directly. That might be on purpose, in which case something like the word "opaque" should appear in the comment explaining this.